pkg/tlsutil/tls.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlsutil
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// CertType defines the type of the cert.
+type CertType int
+
+const (
+	// ClientCert defines a client cert.
+	ClientCert CertType = iota
+	// ServingCert defines a serving cert.
+	ServingCert
+	// ClientAndServingCert defines both client and serving cert.
+	ClientAndServingCert
+)
+
+// CertConfig configures how to generate the Cert.
+type CertConfig struct {
+	// CertName is the name of the cert.
+	CertName string
+	// Optional CertType. Serving, client or both; defaults to both.
+	CertType CertType
+	// Optional CommonName is the common name of the cert; defaults to "".
+	CommonName string
+	// Optional Organization is Organization of the cert; defaults to "".
+	Organization []string
+	// Optional CA Key, if user wants to provide custom CA key via a file path.
+	CAKey string
+	// Optional CA Certificate, if user wants to provide custom CA cert via file path.
+	CACert string
+	// TODO: consider to add passed in SAN fields.
+}
+
+// CertGenerator is an operator specific TLS tool that generates TLS assets for the deploying a user's application.
+type CertGenerator interface {
+	// GenerateCert generates a secret containing TLS encryption key and cert, a Secret
+	// containing the CA key, and a ConfigMap containing the CA Certificate given the Custom
+	// Resource(CR) "cr", the Kubernetes Service "Service", and the CertConfig "config".
+	//
+	// GenerateCert creates and manages TLS key and cert and CA with the following:
+	// CA creation and management:
+	// - If CA is not given:
+	//  - A unique CA is generated for the CR.
+	//  - CA's key is packaged into a Secret as shown below.
+	//  - CA's cert is packaged in a ConfigMap as shown below.
+	//  - The CA Secret and ConfigMap are created on the k8s cluster in the CR's namespace before
+	//    returned to the user. The CertGenerator manages the CA Secret and ConfigMap to ensure it's
+	//    unqiue per CR.
+	// - If CA is given:
+	//  - CA's key is packaged into a Secret as shown below.
+	//  - CA's cert is packaged in a ConfigMap as shown below.
+	//  - The CA Secret and ConfigMap are returned but not created in the K8s cluster in the CR's
```

Comment:
this might be a silly question, but did we consider that these returns are nil? 

I think that it would be less confusing for a user if they just didn't get those return values?

Also with 4 return values should we consider a wrapper struct?

Reply:
>  I might want to mount the CA Cert ConfigMap to a pod so that we can use that to verify the cert that was generated from this CA

Sorry, I am little confused, if the ConfigMap is not created in the cluster because the CA was a custom CA, then this would fail right?

Reply:
@shawn-hurley If the CA is a custom CA passed in via CertConfig then `GenerateCert()` will only package the CA cert up as a Configmap and return that object, but not actually create it on the cluster. So it wouldn't really fail in that case.

In the custom CA case it's up to the users whether or not they should create the CA Configmap and Secret returned by `GenerateCert()` since the operator does not mange the lifecycle of the CA.